core/lib/types/src/kzg.rs

```diff
@@ -0,0 +1,402 @@
+            0u64,
+            0u64,
+        ]))
+        .expect("should have a valid field element from 16 bytes");
+
+        let (opening_proof, opening_value) = compute_proof(&poly, &opening_point_repr);
+
+        let blob_proof = compute_proof_poly(&poly, &kzg_commitment);
+
+        let blob_bytes = zksync_pubdata_into_ethereum_4844_data(&zksync_blob);
+        let mut blob = [0u8; EIP_4844_BYTES_PER_BLOB];
+        blob.copy_from_slice(&blob_bytes);
+
+        let mut commitment = [0u8; 48];
+        commitment.copy_from_slice(kzg_commitment.into_compressed().as_ref());
+
+        let mut challenge_point = [0u8; 32];
+        challenge_point[16..].copy_from_slice(&opening_point.to_be_bytes());
+
+        let mut challenge_value = [0u8; 32];
+        opening_value
+            .into_repr()
+            .write_be(&mut challenge_value[..])
+            .unwrap();
+
+        let mut challenge_proof = [0u8; 48];
+        challenge_proof.copy_from_slice(opening_proof.into_compressed().as_ref());
+
+        let mut commitment_proof = [0u8; 48];
+        commitment_proof.copy_from_slice(blob_proof.into_compressed().as_ref());
+
+        Self {
+            blob,
+            kzg_commitment: commitment,
+            opening_point: challenge_point,
+            opening_value: challenge_value,
+            opening_proof: challenge_proof,
+            versioned_hash,
+            blob_proof: commitment_proof,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use serde::{Deserialize, Serialize};
+    use serde_with::serde_as;
+    use zkevm_circuits::{
+        boojum::pairing::bls12_381::{Fr, FrRepr},
+        eip_4844::{
+            bitreverse, ethereum_4844_data_into_zksync_pubdata, fft,
+            zksync_pubdata_into_monomial_form_poly,
+        },
+    };
+    use zkevm_test_harness::ff::PrimeField;
+    use zkevm_test_harness_1_4_1::{
+        boojum::pairing::{bls12_381::G1Compressed, EncodedPoint},
+        kzg::{verify_kzg_proof, verify_proof_poly},
+    };
+
+    use super::KzgInfo;
+
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize)]
+    struct ExpectedOutputs {
+        #[serde_as(as = "serde_with::hex::Hex")]
+        versioned_hash: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        kzg_commitment: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        opening_point: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        opening_value: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        opening_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        blob_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata_commitment: Vec<u8>,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize)]
+    struct KzgTest {
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata: Vec<u8>,
+        expected_outputs: ExpectedOutputs,
+    }
+
+    fn u8_repr_to_fr(bytes: &[u8]) -> Fr {
```

Comment:
if it is a copy of a function from harness, please add a link to the original function